### PR TITLE
tbtree optimizations

### PR DESCRIPTION
**Pull request by jeroiraz**
This PR introduces optimizations in bulk insertions by following an optimistic concurrent approach.
KV pairs to be inserted are first distributed among corresponding nodes and insertion takes place at the same time, finally the new list of nodes is built up.

Signed-off-by: Jeronimo Irazabal <jeronimo.irazabal@gmail.com>

### CODE REVIEW

File: embedded/tbtree/tbtree.go [hunk 1]
```diff
@@ -1758,63 +1758,102 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
...
+		}
 	}
 
-	copy(newNode.nodes, n.nodes[:insertAt])
-	copy(newNode.nodes[insertAt:], cs)
-	copy(newNode.nodes[insertAt+len(cs):], n.nodes[insertAt+1:])
+	n.nodes = ns
 
-	nodes, err = newNode.split()
+	nodes, _ = n.split()
```
Comment (arriqaaq) [inline embedded/tbtree/tbtree.go:1854]: are we ignoring error here intentionally?

File: embedded/tbtree/tbtree.go [hunk 2]
```diff
@@ -641,7 +640,7 @@ func OpenWith(path string, nLog, hLog, cLog appendable.Appendable, opts *Options
 	}
 
 	if validatedCLogEntry == nil {
-		t.root = &leafNode{t: t, mut: true}
+		t.root = &leafNode{t: t, mut: false}
```
Comment (arriqaaq) [inline embedded/tbtree/tbtree.go:643]: why is this changed to false?

File: embedded/tbtree/tbtree.go (hunk 1 above)
Comment (jeroiraz) [reply embedded/tbtree/tbtree.go:1856]: inner nodes does not return error during split, but I'll add an special error handling

File: embedded/tbtree/tbtree.go (hunk 2 above)
Comment (jeroiraz) [reply embedded/tbtree/tbtree.go:643]: there is just a small difference, if the node start mutated it means changes will be done in the same instance, otherwise a copy will be made. It's better to always do the changes in a copy so any rollback can be done, even when starting with a fresh btree

File: embedded/tbtree/tbtree.go [hunk 3]
```diff
@@ -1760,63 +1766,105 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
...
+		childIndex := n.indexOf(kv.K)
+		kvsPerChild[childIndex] = append(kvsPerChild[childIndex], kv)
 	}
 
-	n._ts = ts
+	var wg sync.WaitGroup
+	wg.Add(len(kvsPerChild))
 
-	ns := make([]node, len(n.nodes)+len(cs)-1)
+	nodesPerChild := make(map[int][]node)
```
Comment (arriqaaq) [inline embedded/tbtree/tbtree.go:1799]: it would be nice if you make this in a struct too, with lock/unlock methods which will make it easier to read/test

File: embedded/tbtree/tbtree.go [hunk 4]
```diff
@@ -1760,63 +1766,105 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
...
 
-	n.nodes = ns
+			c := n.nodes[childIndex]
 
-	nodes, err = n.split()
+			cs, cdepth, cerr := c.insertAt(childKVs, ts)
 
-	return nodes, depth + 1, err
-}
+			nodesMutex.Lock()
```
Comment (arriqaaq) [inline embedded/tbtree/tbtree.go:1811]: shouldn't the lock be acquired after the error?

File: embedded/tbtree/tbtree.go [hunk 5]
```diff
@@ -1760,63 +1766,105 @@ func (t *TBtree) snapshotClosed(snapshot *Snapshot) error {
...
+			defer nodesMutex.Unlock()
 
-func (n *innerNode) copyOnInsertAt(key []byte, value []byte, ts uint64) (nodes []node, depth int, err error) {
-	insertAt := n.indexOf(key)
+			if err != nil {
```
Comment (arriqaaq) [inline embedded/tbtree/tbtree.go:1814]: shouldn't the check be on cerr? Or am I missing something?

Comment (jeroiraz) [reply embedded/tbtree/tbtree.go:1814]: yep

**Review by arriqaaq — APPROVED** [review]
